src/file/properties.rs

```diff
@@ -0,0 +1,550 @@
+    self.default_column_properties.set_encoding(value);
+    self
+  }
+
+  /// Sets compression codec encoding for any column.
+  pub fn set_compression(mut self, value: Compression) -> Self {
+    self.default_column_properties.set_compression(value);
+    self
+  }
+
+  /// Sets flag to enable/disable dictionary encoding for any column.
+  pub fn set_dictionary_enabled(mut self, value: bool) -> Self {
+    self.default_column_properties.set_dictionary_enabled(value);
+    self
+  }
+
+  /// Sets flag to enable/disable statistics for any column.
+  pub fn set_statistics_enabled(mut self, value: bool) -> Self {
+    self.default_column_properties.set_statistics_enabled(value);
+    self
+  }
+
+  /// Sets max statistics size for any column.
+  /// Applicable only if statistics are enabled.
+  pub fn set_max_statistics_size(mut self, value: usize) -> Self {
+    self.default_column_properties.set_max_statistics_size(value);
+    self
+  }
+
+  // ----------------------------------------------------------------------
+  // Setters for a specific column
+
+  /// Helper method to get existing or new mutable reference of column properties.
+  #[inline]
+  fn get_mut_props(&mut self, col: ColumnPath) -> &mut ColumnProperties {
+    self.column_properties.entry(col).or_insert(ColumnProperties::new())
+  }
+
+  /// Sets encoding for a column.
+  /// Takes precedence over globally defined settings.
+  pub fn set_column_encoding(mut self, col: ColumnPath, value: Encoding) -> Self {
+    self.get_mut_props(col).set_encoding(value);
+    self
+  }
+
+  /// Sets compression codec for a column.
+  /// Takes precedence over globally defined settings.
+  pub fn set_column_compression(mut self, col: ColumnPath, value: Compression) -> Self {
+    self.get_mut_props(col).set_compression(value);
+    self
+  }
+
+  /// Sets flag to enable/disable dictionary encoding for a column.
+  /// Takes precedence over globally defined settings.
+  pub fn set_column_dictionary_enabled(mut self, col: ColumnPath, value: bool) -> Self {
+    self.get_mut_props(col).set_dictionary_enabled(value);
+    self
+  }
+
+  /// Sets flag to enable/disable statistics for a column.
+  /// Takes precedence over globally defined settings.
+  pub fn set_column_statistics_enabled(mut self, col: ColumnPath, value: bool) -> Self {
+    self.get_mut_props(col).set_statistics_enabled(value);
+    self
+  }
+
+  /// Sets max size for statistics for a column.
+  /// Takes precedence over globally defined settings.
+  pub fn set_column_max_statistics_size(mut self, col: ColumnPath, value: usize) -> Self {
+    self.get_mut_props(col).set_max_statistics_size(value);
+    self
+  }
+}
+
+/// Container for column properties that can be changed as part of writer.
+///
+/// If a field is `None`, it means that no specific value has been set for this column,
+/// so some default value must be used.
+#[derive(Debug, Clone, PartialEq)]
+struct ColumnProperties {
+  encoding: Option<Encoding>,
```

Comment:
Why these are `Option`s? can they be initialized with the global default settings?

Reply:
I do not think so. The reason is this:

Let's assume that global default compression is UNCOMPRESSED and encoding is PLAIN. I set column A to have encoding RLE, while it has all others as globals, then I set compression for all columns to be SNAPPY, it is not very convenient to propagate this value to the per-column value, e.g. SNAPPY instead of UNCOMPRESSED.

This is how it works now:

For non-column settings, e.g. page size => initially we set them to the global defaults, when value changes, we just set it in a builder and propagate to properties. For column settings, e.g. encoding or compression => per column settings -> global (all columns) settings -> global defaults

For example, I can set global encoding as RLE, but for column A I set DELTA_BINARY_PACKED. So it 
will look like the following:

Column | Per column settings | Global (all columns) settings | Global defaults |
---|---|---|---
Column A | DELTA_BINARY_PACKED | RLE | PLAIN
Column B | | RLE | PLAIN

The reason I have Options in ColumnProperties, so I know whether or not the value has been set and 
I can fall back to the global settings. And because I want to reuse ColumnProperties for per-column 
and default settings, that is why we have another column "Global defaults".

This is how I extract the value:
If per column value is set, extract that value; otherwise, check the global all columns defaults;
if that is not set, return global default.
```rust
self.column_properties.get(col).and_then(|c| c.encoding())
  .or_else(|| self.default_column_properties.encoding())
  .unwrap_or(DEFAULT_ENCODING)
```

But it does look like I can initialise global (all columns) settings to global defaults, so we will have only 2 columns in the table above. I can do something like this:
```rust
self.column_properties.get(col).and_then(|c| c.encoding())
  .or_else(|| self.default_column_properties.encoding())
  .unwrap()
```
So all properties in `self.default_column_properties` will have Some(DEFAULT_XYZ).

Let me know what you think, or how you would do it differently, potentially simpler than this approach. Thanks!

Reply:
Got it. This makes sense to me. Thanks for the explanation!